### PR TITLE
Wip snap misc

### PR DESCRIPTION
**Pull request by ukernel**

### CODE REVIEW

File: src/mds/StrayManager.cc [hunk 1]
```diff
@@ -573,6 +573,17 @@ bool StrayManager::eval_stray(CDentry *dn, bool delay)
...
         }
       }
+
+      if (!in->remote_parents.empty()) {
+	// unlink any stale remote snap dentry.
+	for (compact_set<CDentry*>::iterator p = in->remote_parents.begin();
+	     p != in->remote_parents.end(); ) {
+	  CDentry *remote_dn = *p;
+	  ++p;
+	  assert(remote_dn->last != CEPH_NOSNAP);
```
Comment (gregsfortytwo) [inline src/mds/StrayManager.cc]: What's this assert for?

File: src/mds/StrayManager.cc [hunk 2]
```diff
@@ -573,6 +573,17 @@ bool StrayManager::eval_stray(CDentry *dn, bool delay)
...
+	  remote_dn->unlink_remote(remote_dn->get_linkage());
```
Comment (gregsfortytwo) [inline src/mds/StrayManager.cc]: There are a bunch of prior checks in this function, but I'm not sure which of them are "making sure the stray directory inode is not referenced by any snapshot". Can you point me at it?
The logic in some of these snapshot functions is very obfuscated! :(

File: src/mds/MDCache.cc [hunk 3]
```diff
@@ -7627,6 +7627,12 @@ int MDCache::path_traverse(MDRequestRef& mdr, Message *req, MDSInternalContextBa
   if (cur->state_test(CInode::STATE_PURGING))
     return -ESTALE;
 
+  // make sure snaprealm are open...
+  if (mdr && cur->snaprealm && !cur->snaprealm->open &&
+      !cur->snaprealm->open_parents(_get_waiter(mdr, req, fin))) {
```
Comment (gregsfortytwo) [inline src/mds/MDCache.cc]: Do we really not have a more abstracted way to check this than by reaching inside the SnapRealm's data members directly?
That sort of thing is common in our MDS code but definitely ripe for improvement...

**Comment by gregsfortytwo** [comment]
This set of changes is highlighting our lack of internal interfaces. Do you think it would be a good idea to set up more around SnapRealms so that they have close functions and things, instead of twiddling the "open" bool?
I'd certainly feel more comfortable reviewing changes using those functions than direct access. ;)

File: src/mds/StrayManager.cc (hunk 1 above)
Comment (ukernel) [inline src/mds/StrayManager.cc]: make sure dentry is snap dentry

File: src/mds/StrayManager.cc (hunk 2 above)
Comment (ukernel) [inline src/mds/StrayManager.cc]: in->snaprealm->has_past_parents()

**Comment by gregsfortytwo** [comment]
http://pulpito.ceph.com/gregf-2015-06-03_14:53:22-fs-greg-fs-testing-3---basic-multi/

**Pull request merged**